carbon-design-system/carbon
# feat(upgrade): add package.json update flow to cli

Closes #10190 

This PR adds in an MVP for the Upgrade CLI to support various upgrade strategies. Currently, it supports previewing and writing changes made to a workspace's `package.json` file. In the future, it will also support running migrations for the updates in an upgrade path.


#### Changelog

**New**

- Use `esbuild` to bundle `packages/upgrade` into a single executable `cli.js`
- Split out our CLI into two distinct areas to help with testability: the core CLI layer and **commands**
  - The CLI layer relies on manual testing and should change infrequently as a result, it's difficult to e2e CLI tools (as far as I'm aware) that require input into stdin based on questions
  - The commands (in this case `commands/upgrade.js`) contain all functionality for a command and have integration tests (and can have unit tests)

**Changed**

- Update ESLint to ignore the generated `cli.js` file in `packages/upgrade`
- Update Jest to correctly compile `.cjs` and `.mjs` files

**Removed**

- Remove files that are not needed for the MVP

#### Testing / Reviewing

- Take a look at the overall approach, ask questions if you're unsure how it works, give feedback if something feels off in this MVP
- Verify tests pass as expected (also make sure the tests accurately cover a broad range of cases)
- Try out the CLI locally by:
  - Pulling down this PR
  - Run `cd packages/upgrade`
  - Run `yarn build` to build the CLI
  - Run `cd fixtures/sample-project` to use our sample project fixture
  - Run `../../bin/carbon-upgrade.js --help` to verify the CLI is up-and-running
  - Take a look at the options / capabilities of what's there to see if it makes sense
  - Run `../../bin/carbon-upgrade.js` and choose an upgrade path. Verify the CLI does not change the file and shows a preview of the changes
  - Run `../../bin/carbon-upgrade.js -w` and choose an upgrade path. Verify the CLI edits the file locally (and discard the changes)
